File: sdk/arena-java-sdk/src/main/java/com/github/kubeflow/arena/model/training/VolcanoJobBuilder.java
package com.github.kubeflow.arena.model.training;

import com.github.kubeflow.arena.enums.TrainingJobType;
import com.github.kubeflow.arena.exceptions.ArenaException;
import com.github.kubeflow.arena.model.fields.Field;
import com.github.kubeflow.arena.model.fields.StringField;
import com.github.kubeflow.arena.model.fields.StringListField;

import java.util.ArrayList;
import java.util.Map;

public class VolcanoJobBuilder {
    protected String jobName;

    protected TrainingJobType jobType;

    protected ArrayList<Field> options;

    protected String command;
    public VolcanoJobBuilder() {
        this.jobType = TrainingJobType.VolcanoTrainingJob;
        this.options =  new ArrayList<Field>();
    }

    public VolcanoJobBuilder name(String name) {
        this.options.add(new StringField("--name",name));
        return this;
    }

    public VolcanoJobBuilder queue(String m) {
        this.options.add(new StringField("--queue",m));
        return this;
    }

    public VolcanoJobBuilder minAvailable(String name) {
        this.options.add(new StringField("--min-available",name));
        return this;
    }

    public VolcanoJobBuilder taskCPU(String cpu) {
        this.options.add(new StringField("--task-cpu",cpu));
        return this;
    }

    public VolcanoJobBuilder schedulerName(String name) {
        this.options.add(new StringField("--scheduler-name",name));
        return this;
    }

    public VolcanoJobBuilder taskImages(ArrayList<String> images) {
        this.options.add(new StringListField("--task-images",images));
        return this;
    }

    public VolcanoJobBuilder taskMemory(String memory) {
        this.options.add(new StringField("--task-memory",memory));
        return this;
    }

    public VolcanoJobBuilder taskName(String name) {
        this.options.add(new StringField("--task-name",name));
        return this;
    }

    public VolcanoJobBuilder taskPort(String port) {
        this.options.add(new StringField("--task-port",port));
        return this;
    }


    public VolcanoJobBuilder taskReplicas(int count) {
        this.options.add(new StringField("--task-replicas",String.valueOf(count)));
        return this;
    }

    public TrainingJob build() throws ArenaException {
        ArrayList<String> args = new ArrayList<String>();
        for (int i = 0;i < this.options.size();i++) {
            Field f = this.options.get(i);
            f.validate();
            for(int j = 0;j < f.options().size();j++) {
                args.add(f.options().get(j));
            }
        }
        return new TrainingJob(this.jobName,this.jobType,args,this.command);
    }

    public VolcanoJobBuilder command(String command) {
        this.command = command;
        return this;
    }
}

File: sdk/arena-python-sdk/arenasdk/training/volcano_job_builder.py
#!/usr/bin/env python
from __future__ import annotations
from arenasdk.enums.types import TrainingJobType
from arenasdk.fields.fields import *
from arenasdk.training.job_builder import JobBuilder
from arenasdk.exceptions.arena_exception import ArenaException
from arenasdk.enums.types import ArenaErrorType
from arenasdk.training.job import TrainingJob

class VolcanoJobBuilder(object):
    def __init__(self):
        self._job_type: TrainingJobType = TrainingJobType.VolcanoTrainingJob
        self._job_name = ""
        self._options: List[Field] = list()
        self._command = ""
        
    def build(self) -> TrainingJob:
        args = list()
        try:
            for field in self._options:
                if not isinstance(field,Field):
                    raise ArenaException(ArenaErrorType.Unknown,"the type of option {} is not Field".format(field))
                field.validate()
                for opt in field.options():
                    args.append(opt)
            return TrainingJob(self._job_name,self._job_type,args,self._command)
        except ArenaException as e:
            raise e         
        
    def with_min_available(self,min_available: str) -> VolcanoJobBuilder:
        self._options.append(StringField("--min-available",min_available))
        return self 

    def with_name(self,name: str) -> VolcanoJobBuilder:
        self._options.append(StringField("--name",name))
        return self 

    def with_queue(self,queue: str) -> VolcanoJobBuilder:
        self._options.append(StringField("--queue",queue))
        return self 

    def with_scheduler_name(self,name: str) -> VolcanoJobBuilder:
        self._options.append(StringField("--scheduler-name",name))
        return self 

    def with_task_cpu(self,cpu: str) -> VolcanoJobBuilder:
        self._options.append(StringField("--task-cpu",cpu))
        return self

    def with_task_images(self,images: List[str]) -> VolcanoJobBuilder:
        self._options.append(StringListField("--task-images",images))
        return self
    
    def with_task_memory(self,memory: str) -> VolcanoJobBuilder:
        self._options.append(StringField("--task-memory",memory))
        return self

    def with_task_name(self,name: str) -> VolcanoJobBuilder:
        self._options.append(StringField("--task-name",name))
        return self

    def with_task_port(self,port: int) -> VolcanoJobBuilder:
        self._options.append(StringField("--task-name",str(port)))
        return self

    def with_task_replicas(self,replicas: int) -> VolcanoJobBuilder:
        self._options.append(StringField("--task-replicas",str(replicas)))
        return self
    
    def with_command(self,command: str) -> VolcanoJobBuilder:
        self._command = command
        return self
